python/ray/data/impl/pipeline_executor.py

```diff
@@ -19,7 +19,9 @@ def pipeline_stage(fn: Callable[[], Dataset[T]]) -> Dataset[T]:
     # Force eager evaluation of all blocks in the pipeline stage. This
     # prevents resource deadlocks due to overlapping stage execution (e.g.,
     # task -> actor stage).
-    return fn().fully_executed()
+    # We allow non-lazy (unrecoverable) input blocks to be destroyed here since no
+    # fan-out is allowed in a pipeline.
+    return fn().fully_executed(preserve_input_blocks=True)
```

Comment:
It looks this should be False?

Reply:
Hmm great catch, this should be covered by tests as well. 🤔 I'll fix this and make sure it's covered by tests.